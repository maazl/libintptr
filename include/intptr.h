#ifndef INTPTR_H
#define INTPTR_H

#include <atomic>
#include <type_traits>
#include <cassert>
#include <stdint.h>

namespace mmutil
{

class alignas(32) ref_count
{
	friend class int_ptr_base;
	friend class atomic_int_ptr_base;

	std::atomic<uintptr_t> count;

protected:
	ref_count() : count(0) {}
	~ref_count() {}
public:
	ref_count(const ref_count&) = delete;
	ref_count& operator=(const ref_count&) = delete;

	inline long use_count();
};

class int_ptr_base
{
	friend class atomic_int_ptr_base;
public:
	static_assert(sizeof(uintptr_t) >= 4, "Cannot deal with less than 32 bit platform.");
	static constexpr unsigned  alignment = alignof(ref_count);
	static constexpr uintptr_t counter_mask = alignment - 1U;
	static constexpr uintptr_t pointer_mask = ~counter_mask;

protected:
	ref_count* ptr;

	static void acquire(ref_count* ptr) noexcept { if (ptr) ptr->count += int_ptr_base::alignment; }
	static bool release(ref_count* ptr) noexcept { return ptr && (ptr->count -= int_ptr_base::alignment) == 0; }
	void swap(int_ptr_base&& r) noexcept { std::swap(ptr, r.ptr); }
};

inline long ref_count::use_count()
{	return static_cast<long>(count / int_ptr_base::alignment);
}

template <typename T>
class int_ptr : private int_ptr_base
{
	static_assert(std::is_convertible<T*, ref_count*>::value, "Only types implementing ref_count may be used with int_ptr");
	friend class std::atomic<int_ptr<T>>;

	constexpr explicit int_ptr(ref_count* raw) { ptr = raw; }

public:
	typedef T element_type;

	constexpr int_ptr() noexcept { ptr = nullptr; }
	          int_ptr(T* ptr) noexcept { acquire(this->ptr = ptr); }
	          int_ptr(const int_ptr<T>& r) noexcept { acquire(ptr = r.ptr); }
	constexpr int_ptr(int_ptr<T>&& r) noexcept { ptr = r.ptr; r.ptr = nullptr; }
	          ~int_ptr() { if (release(ptr)) delete &static_cast<T&>(*ptr); }
	// Basic operators
	constexpr T* get() const noexcept { return static_cast<T*>(ptr); }
	constexpr explicit operator bool() const noexcept { return ptr != nullptr; }
	constexpr T& operator*() const { assert(ptr); return *get(); }
	constexpr T* operator->() const { assert(ptr); return get(); }
	friend void  std::swap(int_ptr<T>& l, int_ptr<T>& r) noexcept { l.swap(r); }
	void         reset() { this->~int_ptr(); ptr = nullptr; }
	// assignment
	int_ptr<T>&  operator=(T* ptr) { swap(int_ptr<T>(ptr)); return *this; }
	int_ptr<T>&  operator=(const int_ptr<T>& r) { swap(int_ptr<T>(r)); return *this; }
	int_ptr<T>&  operator=(int_ptr<T>&& r) noexcept { swap(std::move(r)); return *this; }
	// manual resource management for adaption of C libraries.
	constexpr T* toCptr() noexcept { T* ret = get(); ptr = nullptr; return ret; }
	static constexpr int_ptr<T> fromCptr(T* ptr) noexcept { return int_ptr<T>(static_cast<ref_count*>(ptr)); }
};

/// Type erasure part of std::atomic<int_ptr<T>>
class atomic_int_ptr_base
{
protected:
	mutable std::atomic<uintptr_t> ptr;

	ref_count* acquire() const noexcept;
	ref_count* reset() noexcept;
	void swap(int_ptr_base& r) noexcept;

	uintptr_t compare_exchange_weak(int_ptr_base& oldval, const int_ptr_base& newval) noexcept;
	uintptr_t compare_exchange_strong(int_ptr_base& oldval, const int_ptr_base& newval) noexcept;

	constexpr explicit atomic_int_ptr_base(ref_count* ptr = nullptr) noexcept : ptr(reinterpret_cast<uintptr_t>(ptr)) {}

public:
#ifndef NDEBUG
	static std::atomic<unsigned> max_outer_count;
#endif
};

}

namespace std
{

template <typename T>
class atomic<mmutil::int_ptr<T>> : private mmutil::atomic_int_ptr_base
{
public:
	typedef mmutil::int_ptr<T> value_type;

private:
	void swap(value_type&& r) noexcept { mmutil::atomic_int_ptr_base::swap(r); }
	/// finalize CAS operation
	/// @param action control action to perform. The result of one of the
	/// atomic_int_ptr_base compare_exchange functions is expected here.<br/>
	/// If the pointer_mask part is non-zero this object needs to be deleted.
	/// The counter_mask part is the return value.
	bool compare_exchange(uintptr_t action)
	{	delete static_cast<T*>(reinterpret_cast<mmutil::ref_count*>(action & mmutil::int_ptr_base::pointer_mask));
		return static_cast<bool>(action & mmutil::int_ptr_base::counter_mask);
	}

public:
	constexpr atomic() noexcept {}
	          atomic(const value_type& r) noexcept : atomic_int_ptr_base(r.ptr) { mmutil::int_ptr_base::acquire(r.ptr); }
	          atomic(const atomic<value_type>& r) = delete;
	constexpr atomic(value_type&& r) noexcept : atomic_int_ptr_base(r.ptr) { r.ptr = nullptr; }
	          ~atomic()
	{	uintptr_t raw = ptr;
		assert((raw & mmutil::int_ptr_base::counter_mask) == 0);
		if (mmutil::int_ptr_base::release(reinterpret_cast<mmutil::ref_count*>(raw)))
			delete &static_cast<T&>(*reinterpret_cast<mmutil::ref_count*>(raw));
	}
	// basic operators
	operator  value_type() const noexcept { return value_type(acquire()); }
	constexpr explicit operator bool() const noexcept { return (ptr & value_type::pointer_mask) != 0; }
	// mutable operators
	void      reset() { delete static_cast<T*>(mmutil::atomic_int_ptr_base::reset()); }
	inline friend void swap(value_type& l, atomic<value_type>& r) noexcept { r.swap(std::move(l)); }
	inline friend void swap(atomic<value_type>& l, value_type& r) noexcept { l.swap(std::move(r)); }
	// assignment
	atomic<value_type>& operator=(const atomic<value_type>&) = delete;
	void      operator=(const value_type& r) { swap(value_type(r)); }
	void      operator=(value_type&& r) noexcept { swap(std::move(r)); }

	bool compare_exchange_weak(value_type& oldval, const value_type& newval)
	{	return compare_exchange(mmutil::atomic_int_ptr_base::compare_exchange_weak(oldval, newval)); }

	bool compare_exchange_strong(value_type& oldval, const value_type& newval)
	{	return compare_exchange(mmutil::atomic_int_ptr_base::compare_exchange_strong(oldval, newval)); }

	bool is_lock_free() const noexcept { return ptr.is_lock_free(); }
#if __cplusplus >= 201703L
	static constexpr bool is_always_lock_free = atomic<uintptr_t>::is_always_lock_free;
#endif
};

}

namespace mmutil
{

template <typename T> using atomic_int_ptr = std::atomic<int_ptr<T>>;

}
#endif
