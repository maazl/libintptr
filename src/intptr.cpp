#include <intptr.h>

using namespace mmutil;

#ifndef NDEBUG
std::atomic<unsigned> atomic_int_ptr_base::max_outer_count(1);
#endif

ref_count* atomic_int_ptr_base::acquire() const noexcept
{	__builtin_prefetch(reinterpret_cast<ref_count*>(ptr & int_ptr_base::pointer_mask), 1);
	uintptr_t old_outer = ++ptr;
	const uintptr_t outer_count = old_outer & int_ptr_base::counter_mask;
	assert(outer_count != 0); // overflow condition
	ref_count* const new_outer = reinterpret_cast<ref_count*>(old_outer & int_ptr_base::pointer_mask);
	if (new_outer)
		// Transfer counter to obj->count.
		new_outer->count += int_ptr_base::alignment + 1 - outer_count;
	// And reset it in *this.
	if (!ptr.compare_exchange_strong(old_outer, reinterpret_cast<uintptr_t>(new_outer)) && new_outer)
		// Someone else does the job already => undo.
		new_outer->count += outer_count;
		// The global count cannot return to zero here, because we have an active reference.
	#ifndef NDEBUG
	unsigned max_outer = max_outer_count;
	while (max_outer < outer_count
		&& !max_outer_count.compare_exchange_weak(max_outer, static_cast<unsigned>(outer_count)));
	#endif
	return new_outer;
}

ref_count* atomic_int_ptr_base::reset() noexcept
{	uintptr_t old = ptr.exchange(0U);
	ref_count* obj = reinterpret_cast<ref_count*>(old & int_ptr_base::pointer_mask);
	return obj && (obj->count -= (old & int_ptr_base::counter_mask) + int_ptr_base::alignment) == 0
		? obj : nullptr;
}

void atomic_int_ptr_base::swap(int_ptr_base& r) noexcept
{ uintptr_t old = ptr.exchange(reinterpret_cast<uintptr_t>(r.ptr));
  // Transfer hold count to the main counter and get the data with hold count 0 into r.
	uintptr_t outer = old & int_ptr_base::counter_mask;
	if (outer && (old &= int_ptr_base::pointer_mask) != 0U)
		reinterpret_cast<ref_count*>(old)->count -= outer;
	r.ptr = reinterpret_cast<ref_count*>(old);
}

uintptr_t atomic_int_ptr_base::compare_exchange_weak(int_ptr_base& oldval, const int_ptr_base& newval) noexcept
{	uintptr_t preval = reinterpret_cast<uintptr_t>(oldval.ptr);
	if (!ptr.compare_exchange_weak(preval, reinterpret_cast<uintptr_t>(newval.ptr)))
	{	preval = reinterpret_cast<uintptr_t>(oldval.ptr);
		oldval.ptr = acquire();
		return preval; // failed
	}
	// success => update reference counts.
	int_ptr_base::acquire(newval.ptr);
	return int_ptr_base::release(oldval.ptr) * preval + 1;
}

uintptr_t atomic_int_ptr_base::compare_exchange_strong(int_ptr_base& oldval, const int_ptr_base& newval) noexcept
{	uintptr_t preval = reinterpret_cast<uintptr_t>(oldval.ptr);
	while (!ptr.compare_exchange_weak(preval, reinterpret_cast<uintptr_t>(newval.ptr)))
		if ((preval ^ reinterpret_cast<uintptr_t>(oldval.ptr)) > int_ptr_base::counter_mask) // compare pointer part only
		{	preval = reinterpret_cast<uintptr_t>(oldval.ptr);
			ref_count* curptr = acquire();
			if (curptr != oldval.ptr)
			{	oldval.ptr = curptr;
				return preval;
			}
			// retrieved pointer happens to match oldval now => undo acquire and retry
			bool del = int_ptr_base::release(curptr);
			// since there is at least one additional strong reference in oldval the release operation must not cause count 0.
			assert(!del);
		}
	// success => update reference counts.
	int_ptr_base::acquire(newval.ptr);
	// release oldval and transfer the outer count if any
	return (oldval.ptr &&
			(oldval.ptr->count -= int_ptr_base::alignment + (preval & int_ptr_base::counter_mask)) == 0)
		* reinterpret_cast<uintptr_t>(oldval.ptr) + 1;
}
