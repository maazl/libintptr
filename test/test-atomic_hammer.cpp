#include <intptr.h>
#include <cstdio>
#include <thread>
#include <chrono>

using namespace mmutil;
using namespace std;
using namespace std::chrono_literals;

static unsigned thread_count = 300;
static auto test_time = 3s;

struct logger
{
#ifndef NDEBUG
	unsigned long value = 0;
	logger() : value(0) {}
	void operator++() { ++value; }
	void print(const char* task, int id) { printf("%s %i: count %lu\n", task, id, value); }
#else
	void operator++() {}
	void print(const char*, int) {}
#endif
};

struct Obj : public ref_count
{ int I;
	Obj(int i) : I(i) { printf("TestObj(%i)::TestObj()\n", I); }
	~Obj() { printf("TestObj(%i)::~TestObj()\n", I); }
};

static bool terminate_test = false;
static atomic<int_ptr<Obj>> instance;

static void writer(int id)
{
	int_ptr<Obj> myObj = new Obj(id);

	logger count;
	do
	{	instance = myObj;
		++count;
	} while (!terminate_test);

	count.print("writer", id);
}

static void reader(int id)
{
	int_ptr<Obj> myObj;

	logger count;
	do
	{	myObj = instance;
		++count;
	} while (!terminate_test);

	count.print("reader", id);
}

static void readwrite(int id)
{
	int_ptr<Obj> myObj = new Obj(id);

	logger count;
	do
	{	instance = myObj;
		myObj = instance;
		++count;
	} while (!terminate_test);

	count.print("readwrite", id);
}

int main()
{
	std::thread threads[thread_count];

	void (*funcs[3])(int) =
	{	writer,
		reader,
		readwrite
	};

	for (unsigned id = 0; id < thread_count; ++id)
		threads[id] = thread(funcs[id%3], id);

	this_thread::sleep_for(test_time);
	terminate_test = true;

	for (unsigned id = 0; id < thread_count; ++id)
		threads[id].join();

	#ifndef NDEBUG
	printf ("max_outer_count: %u\n", atomic_int_ptr_base::max_outer_count.load());
	#endif

	return 0;
}
