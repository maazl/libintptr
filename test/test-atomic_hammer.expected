TestObj(0)::~TestObj()
TestObj(0)::TestObj()
TestObj(101)::~TestObj()
TestObj(101)::TestObj()
TestObj(102)::~TestObj()
TestObj(102)::TestObj()
TestObj(104)::~TestObj()
TestObj(104)::TestObj()
TestObj(105)::~TestObj()
TestObj(105)::TestObj()
TestObj(107)::~TestObj()
TestObj(107)::TestObj()
TestObj(108)::~TestObj()
TestObj(108)::TestObj()
TestObj(110)::~TestObj()
TestObj(110)::TestObj()
TestObj(111)::~TestObj()
TestObj(111)::TestObj()
TestObj(113)::~TestObj()
TestObj(113)::TestObj()
TestObj(114)::~TestObj()
TestObj(114)::TestObj()
TestObj(116)::~TestObj()
TestObj(116)::TestObj()
TestObj(117)::~TestObj()
TestObj(117)::TestObj()
TestObj(119)::~TestObj()
TestObj(119)::TestObj()
TestObj(11)::~TestObj()
TestObj(11)::TestObj()
TestObj(120)::~TestObj()
TestObj(120)::TestObj()
TestObj(122)::~TestObj()
TestObj(122)::TestObj()
TestObj(123)::~TestObj()
TestObj(123)::TestObj()
TestObj(125)::~TestObj()
TestObj(125)::TestObj()
TestObj(126)::~TestObj()
TestObj(126)::TestObj()
TestObj(128)::~TestObj()
TestObj(128)::TestObj()
TestObj(129)::~TestObj()
TestObj(129)::TestObj()
TestObj(12)::~TestObj()
TestObj(12)::TestObj()
TestObj(131)::~TestObj()
TestObj(131)::TestObj()
TestObj(132)::~TestObj()
TestObj(132)::TestObj()
TestObj(134)::~TestObj()
TestObj(134)::TestObj()
TestObj(135)::~TestObj()
TestObj(135)::TestObj()
TestObj(137)::~TestObj()
TestObj(137)::TestObj()
TestObj(138)::~TestObj()
TestObj(138)::TestObj()
TestObj(140)::~TestObj()
TestObj(140)::TestObj()
TestObj(141)::~TestObj()
TestObj(141)::TestObj()
TestObj(143)::~TestObj()
TestObj(143)::TestObj()
TestObj(144)::~TestObj()
TestObj(144)::TestObj()
TestObj(146)::~TestObj()
TestObj(146)::TestObj()
TestObj(147)::~TestObj()
TestObj(147)::TestObj()
TestObj(149)::~TestObj()
TestObj(149)::TestObj()
TestObj(14)::~TestObj()
TestObj(14)::TestObj()
TestObj(150)::~TestObj()
TestObj(150)::TestObj()
TestObj(152)::~TestObj()
TestObj(152)::TestObj()
TestObj(153)::~TestObj()
TestObj(153)::TestObj()
TestObj(155)::~TestObj()
TestObj(155)::TestObj()
TestObj(156)::~TestObj()
TestObj(156)::TestObj()
TestObj(158)::~TestObj()
TestObj(158)::TestObj()
TestObj(159)::~TestObj()
TestObj(159)::TestObj()
TestObj(15)::~TestObj()
TestObj(15)::TestObj()
TestObj(161)::~TestObj()
TestObj(161)::TestObj()
TestObj(162)::~TestObj()
TestObj(162)::TestObj()
TestObj(164)::~TestObj()
TestObj(164)::TestObj()
TestObj(165)::~TestObj()
TestObj(165)::TestObj()
TestObj(167)::~TestObj()
TestObj(167)::TestObj()
TestObj(168)::~TestObj()
TestObj(168)::TestObj()
TestObj(170)::~TestObj()
TestObj(170)::TestObj()
TestObj(171)::~TestObj()
TestObj(171)::TestObj()
TestObj(173)::~TestObj()
TestObj(173)::TestObj()
TestObj(174)::~TestObj()
TestObj(174)::TestObj()
TestObj(176)::~TestObj()
TestObj(176)::TestObj()
TestObj(177)::~TestObj()
TestObj(177)::TestObj()
TestObj(179)::~TestObj()
TestObj(179)::TestObj()
TestObj(17)::~TestObj()
TestObj(17)::TestObj()
TestObj(180)::~TestObj()
TestObj(180)::TestObj()
TestObj(182)::~TestObj()
TestObj(182)::TestObj()
TestObj(183)::~TestObj()
TestObj(183)::TestObj()
TestObj(185)::~TestObj()
TestObj(185)::TestObj()
TestObj(186)::~TestObj()
TestObj(186)::TestObj()
TestObj(188)::~TestObj()
TestObj(188)::TestObj()
TestObj(189)::~TestObj()
TestObj(189)::TestObj()
TestObj(18)::~TestObj()
TestObj(18)::TestObj()
TestObj(191)::~TestObj()
TestObj(191)::TestObj()
TestObj(192)::~TestObj()
TestObj(192)::TestObj()
TestObj(194)::~TestObj()
TestObj(194)::TestObj()
TestObj(195)::~TestObj()
TestObj(195)::TestObj()
TestObj(197)::~TestObj()
TestObj(197)::TestObj()
TestObj(198)::~TestObj()
TestObj(198)::TestObj()
TestObj(200)::~TestObj()
TestObj(200)::TestObj()
TestObj(201)::~TestObj()
TestObj(201)::TestObj()
TestObj(203)::~TestObj()
TestObj(203)::TestObj()
TestObj(204)::~TestObj()
TestObj(204)::TestObj()
TestObj(206)::~TestObj()
TestObj(206)::TestObj()
TestObj(207)::~TestObj()
TestObj(207)::TestObj()
TestObj(209)::~TestObj()
TestObj(209)::TestObj()
TestObj(20)::~TestObj()
TestObj(20)::TestObj()
TestObj(210)::~TestObj()
TestObj(210)::TestObj()
TestObj(212)::~TestObj()
TestObj(212)::TestObj()
TestObj(213)::~TestObj()
TestObj(213)::TestObj()
TestObj(215)::~TestObj()
TestObj(215)::TestObj()
TestObj(216)::~TestObj()
TestObj(216)::TestObj()
TestObj(218)::~TestObj()
TestObj(218)::TestObj()
TestObj(219)::~TestObj()
TestObj(219)::TestObj()
TestObj(21)::~TestObj()
TestObj(21)::TestObj()
TestObj(221)::~TestObj()
TestObj(221)::TestObj()
TestObj(222)::~TestObj()
TestObj(222)::TestObj()
TestObj(224)::~TestObj()
TestObj(224)::TestObj()
TestObj(225)::~TestObj()
TestObj(225)::TestObj()
TestObj(227)::~TestObj()
TestObj(227)::TestObj()
TestObj(228)::~TestObj()
TestObj(228)::TestObj()
TestObj(230)::~TestObj()
TestObj(230)::TestObj()
TestObj(231)::~TestObj()
TestObj(231)::TestObj()
TestObj(233)::~TestObj()
TestObj(233)::TestObj()
TestObj(234)::~TestObj()
TestObj(234)::TestObj()
TestObj(236)::~TestObj()
TestObj(236)::TestObj()
TestObj(237)::~TestObj()
TestObj(237)::TestObj()
TestObj(239)::~TestObj()
TestObj(239)::TestObj()
TestObj(23)::~TestObj()
TestObj(23)::TestObj()
TestObj(240)::~TestObj()
TestObj(240)::TestObj()
TestObj(242)::~TestObj()
TestObj(242)::TestObj()
TestObj(243)::~TestObj()
TestObj(243)::TestObj()
TestObj(245)::~TestObj()
TestObj(245)::TestObj()
TestObj(246)::~TestObj()
TestObj(246)::TestObj()
TestObj(248)::~TestObj()
TestObj(248)::TestObj()
TestObj(249)::~TestObj()
TestObj(249)::TestObj()
TestObj(24)::~TestObj()
TestObj(24)::TestObj()
TestObj(251)::~TestObj()
TestObj(251)::TestObj()
TestObj(252)::~TestObj()
TestObj(252)::TestObj()
TestObj(254)::~TestObj()
TestObj(254)::TestObj()
TestObj(255)::~TestObj()
TestObj(255)::TestObj()
TestObj(257)::~TestObj()
TestObj(257)::TestObj()
TestObj(258)::~TestObj()
TestObj(258)::TestObj()
TestObj(260)::~TestObj()
TestObj(260)::TestObj()
TestObj(261)::~TestObj()
TestObj(261)::TestObj()
TestObj(263)::~TestObj()
TestObj(263)::TestObj()
TestObj(264)::~TestObj()
TestObj(264)::TestObj()
TestObj(266)::~TestObj()
TestObj(266)::TestObj()
TestObj(267)::~TestObj()
TestObj(267)::TestObj()
TestObj(269)::~TestObj()
TestObj(269)::TestObj()
TestObj(26)::~TestObj()
TestObj(26)::TestObj()
TestObj(270)::~TestObj()
TestObj(270)::TestObj()
TestObj(272)::~TestObj()
TestObj(272)::TestObj()
TestObj(273)::~TestObj()
TestObj(273)::TestObj()
TestObj(275)::~TestObj()
TestObj(275)::TestObj()
TestObj(276)::~TestObj()
TestObj(276)::TestObj()
TestObj(278)::~TestObj()
TestObj(278)::TestObj()
TestObj(279)::~TestObj()
TestObj(279)::TestObj()
TestObj(27)::~TestObj()
TestObj(27)::TestObj()
TestObj(281)::~TestObj()
TestObj(281)::TestObj()
TestObj(282)::~TestObj()
TestObj(282)::TestObj()
TestObj(284)::~TestObj()
TestObj(284)::TestObj()
TestObj(285)::~TestObj()
TestObj(285)::TestObj()
TestObj(287)::~TestObj()
TestObj(287)::TestObj()
TestObj(288)::~TestObj()
TestObj(288)::TestObj()
TestObj(290)::~TestObj()
TestObj(290)::TestObj()
TestObj(291)::~TestObj()
TestObj(291)::TestObj()
TestObj(293)::~TestObj()
TestObj(293)::TestObj()
TestObj(294)::~TestObj()
TestObj(294)::TestObj()
TestObj(296)::~TestObj()
TestObj(296)::TestObj()
TestObj(297)::~TestObj()
TestObj(297)::TestObj()
TestObj(299)::~TestObj()
TestObj(299)::TestObj()
TestObj(29)::~TestObj()
TestObj(29)::TestObj()
TestObj(2)::~TestObj()
TestObj(2)::TestObj()
TestObj(30)::~TestObj()
TestObj(30)::TestObj()
TestObj(32)::~TestObj()
TestObj(32)::TestObj()
TestObj(33)::~TestObj()
TestObj(33)::TestObj()
TestObj(35)::~TestObj()
TestObj(35)::TestObj()
TestObj(36)::~TestObj()
TestObj(36)::TestObj()
TestObj(38)::~TestObj()
TestObj(38)::TestObj()
TestObj(39)::~TestObj()
TestObj(39)::TestObj()
TestObj(3)::~TestObj()
TestObj(3)::TestObj()
TestObj(41)::~TestObj()
TestObj(41)::TestObj()
TestObj(42)::~TestObj()
TestObj(42)::TestObj()
TestObj(44)::~TestObj()
TestObj(44)::TestObj()
TestObj(45)::~TestObj()
TestObj(45)::TestObj()
TestObj(47)::~TestObj()
TestObj(47)::TestObj()
TestObj(48)::~TestObj()
TestObj(48)::TestObj()
TestObj(50)::~TestObj()
TestObj(50)::TestObj()
TestObj(51)::~TestObj()
TestObj(51)::TestObj()
TestObj(53)::~TestObj()
TestObj(53)::TestObj()
TestObj(54)::~TestObj()
TestObj(54)::TestObj()
TestObj(56)::~TestObj()
TestObj(56)::TestObj()
TestObj(57)::~TestObj()
TestObj(57)::TestObj()
TestObj(59)::~TestObj()
TestObj(59)::TestObj()
TestObj(5)::~TestObj()
TestObj(5)::TestObj()
TestObj(60)::~TestObj()
TestObj(60)::TestObj()
TestObj(62)::~TestObj()
TestObj(62)::TestObj()
TestObj(63)::~TestObj()
TestObj(63)::TestObj()
TestObj(65)::~TestObj()
TestObj(65)::TestObj()
TestObj(66)::~TestObj()
TestObj(66)::TestObj()
TestObj(68)::~TestObj()
TestObj(68)::TestObj()
TestObj(69)::~TestObj()
TestObj(69)::TestObj()
TestObj(6)::~TestObj()
TestObj(6)::TestObj()
TestObj(71)::~TestObj()
TestObj(71)::TestObj()
TestObj(72)::~TestObj()
TestObj(72)::TestObj()
TestObj(74)::~TestObj()
TestObj(74)::TestObj()
TestObj(75)::~TestObj()
TestObj(75)::TestObj()
TestObj(77)::~TestObj()
TestObj(77)::TestObj()
TestObj(78)::~TestObj()
TestObj(78)::TestObj()
TestObj(80)::~TestObj()
TestObj(80)::TestObj()
TestObj(81)::~TestObj()
TestObj(81)::TestObj()
TestObj(83)::~TestObj()
TestObj(83)::TestObj()
TestObj(84)::~TestObj()
TestObj(84)::TestObj()
TestObj(86)::~TestObj()
TestObj(86)::TestObj()
TestObj(87)::~TestObj()
TestObj(87)::TestObj()
TestObj(89)::~TestObj()
TestObj(89)::TestObj()
TestObj(8)::~TestObj()
TestObj(8)::TestObj()
TestObj(90)::~TestObj()
TestObj(90)::TestObj()
TestObj(92)::~TestObj()
TestObj(92)::TestObj()
TestObj(93)::~TestObj()
TestObj(93)::TestObj()
TestObj(95)::~TestObj()
TestObj(95)::TestObj()
TestObj(96)::~TestObj()
TestObj(96)::TestObj()
TestObj(98)::~TestObj()
TestObj(98)::TestObj()
TestObj(99)::~TestObj()
TestObj(99)::TestObj()
TestObj(9)::~TestObj()
TestObj(9)::TestObj()
